### PR TITLE
Optimization + Threaded

### PR DESCRIPTION
**Pull request by niyonx**
## Optimization Changes

- Thread-safe multithreading
- Expectedrows estimation for parsing
- Python best practices
- In-kernel Pytables search
- Code changes

### CODE REVIEW

File: pytracer/gui/callbacks.py [hunk 1]
```diff
@@ -18,45 +18,50 @@
...
-    b = time.perf_counter()
+    # b = time.perf_counter()
     rows = pgc.get_active_row(selected_rows, data)
     rows_str = [
         f"module: {d['module']}, function: {d['function']}" for d in rows]
     msg = f"Selected rows:\n {os.linesep.join(rows_str)}"
-    e = time.perf_counter()
-    print("update_table_active_cell", e-b)
+    # e = time.perf_counter()
+    # print("update_table_active_cell", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:56]: Remove comments.

File: pytracer/gui/callbacks.py [hunk 2]
```diff
@@ -179,12 +187,12 @@ def print_heatmap(hover_data, mode, color, zscale, fig, lstart, lend):
             fig.update_layout(coloraxis=colorscale)
             figure = fig
 
-    e = time.perf_counter()
-    print("print_heatmap", e-b)
+    # e = time.perf_counter()
+    # print("print_heatmap", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:191]: Remove comments

File: pytracer/gui/callbacks.py [hunk 3]
```diff
@@ -231,10 +239,11 @@ def get_full_source_line(path, line):
     dash.dependencies.Output("source", "children"),
     dash.dependencies.Output("source-link", "href"),
     dash.dependencies.Output("source-link", "children"),
-    [dash.dependencies.Input("timeline", "hoverData")])
+    [dash.dependencies.Input("timeline", "hoverData")],
+    prevent_initial_call=True)
 # @cache.memoize(timeout=TIMEOUT)
 def print_source(hover_data):
-    b = time.perf_counter()
+    # b = time.perf_counter()
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:246]: Remove comments

File: pytracer/gui/callbacks.py [hunk 4]
```diff
@@ -252,8 +261,8 @@ def print_source(hover_data):
             raise FileNotFoundError
         description = f"{source}:{_lineno}"
 
-    e = time.perf_counter()
-    print("print_source", e-b)
+    # e = time.perf_counter()
+    # print("print_source", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:265]: Remove comments

File: pytracer/gui/callbacks.py [hunk 5]
```diff
@@ -306,26 +316,28 @@ def print_modal_source(on, href, href_description):
                                         annotations=[{'row': line_start-1,
                                                       'type': 'error', 'text': 'Current call'}])
 
-    e = time.perf_counter()
-    print("print_modal_source", e-b)
+    # e = time.perf_counter()
+    # print("print_modal_source", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:320]: Remove comments

File: pytracer/gui/callbacks.py [hunk 6]
```diff
@@ -372,45 +385,48 @@ def print_datahover_summary(hover_data, fig, mode):
                         f"shape={shape}{os.linesep}{os.linesep}"
                         f"Frobenius norm={norm_fro:.2}{os.linesep}{os.linesep}")
 
-    e = time.perf_counter()
-    print("print_datahover_summary", e-b)
+    # e = time.perf_counter()
+    # print("print_datahover_summary", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:389]: Remove comments

File: pytracer/gui/callbacks.py [hunk 7]
```diff
@@ -372,45 +385,48 @@ def print_datahover_summary(hover_data, fig, mode):
...
     dash.dependencies.Input("source-button", "on"))
 def open_modal_source(on):
-    b = time.perf_counter()
+    # b = time.perf_counter()
     style_off = {"display": "none"}
     style_on = {"display": "block", "width": "100%", "height": 300}
-    e = time.perf_counter()
-    print("open_modal_source", e-b)
+    # e = time.perf_counter()
+    # print("open_modal_source", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:401]: Remove comments

File: pytracer/gui/callbacks.py [hunk 8]
```diff
@@ -372,45 +385,48 @@ def print_datahover_summary(hover_data, fig, mode):
...
-            row["name"] == arg and \
-            row["BacktraceDescription"] == backtrace and \
-            row['time'] >= time_start and \
-            row['time'] <= time_end
-
-    x = pgc.data.filter(module, function, get_x, "time")
-    y = pgc.data.filter(module, function, get_x, mode)
+    # b = time.perf_counter()
+
+    # b1 = time.perf_counter()
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:410]: Remove comments

File: pytracer/gui/callbacks.py [hunk 9]
```diff
@@ -372,45 +385,48 @@ def print_datahover_summary(hover_data, fig, mode):
...
+    y = pgc.data.filter(module, function, get_x, mode, arg, label)
     (filename, line, lineno, name) = backtrace
 
-    e1 = time.perf_counter()
-    print("pgc.data.filter x,y", e1-b1)
+    # e1 = time.perf_counter()
+    # print("pgc.data.filter x,y", e1-b1)
 
-    b2 = time.perf_counter()
+    # b2 = time.perf_counter()
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:429]: Remove comments.

File: pytracer/gui/callbacks.py [hunk 10]
```diff
@@ -420,17 +436,19 @@ def get_x(row):
...
+        customdata_append(info)
+        hovertext_append(f"{function}{os.linesep}{arg.decode('utf-8')}")
 
-    e2 = time.perf_counter()
-    print("extra_value", e2-b2)
+    # e2 = time.perf_counter()
+    # print("extra_value", e2-b2)
 
-    b3 = time.perf_counter()
+    # b3 = time.perf_counter()
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:451]: Remove comments

File: pytracer/gui/callbacks.py [hunk 11]
```diff
@@ -443,19 +461,19 @@ def get_x(row):
...
                            marker_color=marker_color)
-    e3 = time.perf_counter()
-    print("scattergl", b3-e3)
+    # e3 = time.perf_counter()
+    # print("scattergl", b3-e3)
 
-    e = time.perf_counter()
-    print("get_scatter_timeline", e-b)
+    # e = time.perf_counter()
+    # print("get_scatter_timeline", e-b)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:468]: Remove comments.

File: pytracer/gui/callbacks.py [hunk 12]
```diff
@@ -471,13 +489,12 @@ def add_scatter(fig, module, function,
...
-                fig.add_trace(scatter)
-
-    e = time.perf_counter()
-    print("add_scatter", e-b)
+            fig.add_trace(scatter)
 
+    # e = time.perf_counter()
+    # print("add_scatter", e-b)
 
+# @cache.memoize(timeout=TIMEOUT)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:497]: Remove comments

File: pytracer/gui/callbacks.py [hunk 13]
```diff
@@ -552,33 +570,44 @@ def update_timeline(selected_rows, data, mode, xscale, yscale,
...
                      rangemode="tozero", type=yscale)
 
-    module_and_function = [data[selected_row]
-                           for selected_row in selected_rows]
+    module_and_function = [*map(lambda x: data[x], selected_rows)]
 
     if line_on:
         if os.path.isfile(lfile):
-            calls = get_first_call_from_line(lfile, lstart)
+            # calls = get_first_call_from_line(lfile, lstart)
```
Comment (yohanchatelain) [inline pytracer/gui/callbacks.py:577]: Remove comments

**Review by yohanchatelain — CHANGES_REQUESTED** [review]
I add small comments but it's great work overall Nigel, congrats! :tada:

File: pytracer/core/parser.py [hunk 14]
```diff
@@ -130,9 +135,9 @@ def merge_dict(self, args):
     def _merge(self, values, attr, do_not_check=False):
         attrs = None
         if isinstance(attr, str):
-            attrs = [value[attr] for value in values]
+            attrs = [*map(lambda value: value[attr], values)]
```
[the remark this answers is not in the document]
Comment (niyonx) [reply pytracer/core/parser.py:138]: Just in general map is faster than for loops. You're right though, list comprehension are faster than map, if it requires a lambda. I will revert the map with lambda back to list comprehension!

File: pytracer/gui/layout.py [hunk 15]
```diff
@@ -11,6 +11,8 @@
 import pytracer.gui.core as pgc
 import pytracer.callgraph.core as pcc
 
+import numpy as np
```
[the remark this answers is not in the document]
Comment (niyonx) [reply pytracer/gui/layout.py:14]: I used np.arange for the start_tiime and end_tiime time generation. I tried multiple approaches and the numpy one was the fastest.

**Comment by niyonx** [comment]
#17 #18 can be linked as well

**Review by yohanchatelain — APPROVED** [review]

**Comment by niyonx** [comment]
> #17 #18 can be linked as well

@yohanchatelain is it good to be merged now?

**Pull request merged**